assets/src/edit-story/utils/useDoubleClick.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useEffect, useState } from '@wordpress/element';
+
+// Disable reason: Should not check indentation in code samples inside
+// markdown, but it does.
+
+/* eslint-disable jsdoc/check-indentation */
+
+/**
+ * This hook creates a handler to use for double click listening
+ * on a node, where single clicks are also relevant. Default timeout
+ * is 200 ms. If no subsequent click has been recorded in the given
+ * time, single click will be assumed.
+ *
+ * This hook returns a handler retrieval function, not the handler
+ * itself. This allows the component to use the same hook for multiple
+ * children listening for events.
+ *
+ * For example:
+ * ```
+ * function SomeComponent() {
+ *   const handleSingle = useCallback( ( evt, item ) => {
+ *     // Handle single click on `item`
+ *   }, [] );
+ *   const handleDouble = useCallback( ( evt, item ) => {
+ *     // Handle double click on `item`
+ *   }, [] );
+ *   const getHandler = useDoubleClick( handleSingle, handleDouble );
+ *   const items = [
+ *     // Some objects with `id` and `name` attributes.
+ *   ];
+ *   return items.map( ( item} ) => (
+ *     <button key={ item.id } onClick={ getHandler( item ) }>
+ *       { item.name }
+ *     </button>
+ *   ) );
+ * }
+ * ```
+ *
+ * @param {Function} onSingleClick  Handler to activate on single click.
+ * @param {Function} onDoubleClick  Handler to activate on double click.
+ * @param {number}   ms             Timeout in ms to wait - defaults to 200.
+ * @return {Function} Handler retrieval function to get an onClick listener (invoke with unique value).
+ */
+const useDoubleClick = ( onSingleClick, onDoubleClick, ms = null ) => {
+	const [ target, setTarget ] = useState( null );
+	const [ lastEvent, setLastEvent ] = useState( null );
+	const getHandler = useCallback( ( newTarget ) => ( evt ) => {
+		evt.stopPropagation();
+
+		if ( target !== newTarget ) {
```

Comment:
I guess it's intended that in case of an image, it first has to be selected and only then double click is required to go to edit mode?

Reply:
It's actually because I wanted this handler to be able to be used in a component will multiple children, like [in the example given in the description of the hook](https://github.com/ampproject/amp-wp/pull/3804/files/575fccf668fbff8c6a004ea45180ac9f9ee9fb4e#diff-55be7466fec1b29684bc0dcb019ca439R23-R39).

If it just counted clicks, it would count a click on one item and then another fast click on another as a double click.

As you can't use a variable number of hooks in a component (it always has to be the exact same amount), this was the best way to create it.